In [166]:
import pandas as pd
df=pd.read_csv("C:\projects\Healthcare-Analytics-Dashboard\python\data\patients.csv")

In [167]:
df

,Id,BIRTHDATE,DEATHDATE,SSN,DRIVERS,PASSPORT,PREFIX,FIRST,LAST,SUFFIX,...,BIRTHPLACE,ADDRESS,CITY,STATE,COUNTY,ZIP,LAT,LON,HEALTHCARE_EXPENSES,HEALTHCARE_COVERAGE
0,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,2017-08-24,NaN,999-68-6630,NaN,NaN,NaN,Jacinto644,Kris249,NaN,...,Beverly Massachusetts US,888 Hickle Ferry Suite 38,Springfield,Massachusetts,Hampden County,1106.0,42.151961,-72.598959,8446.49,1499.08
1,067318a4-db8f-447f-8b6e-f2f61e9baaa5,2016-08-01,NaN,999-15-5895,NaN,NaN,NaN,Alva958,Krajcik437,NaN,...,Boston Massachusetts US,1048 Skiles Trailer,Walpole,Massachusetts,Norfolk County,2081.0,42.177370,-71.281353,89893.40,1845.72
2,ae9efba3-ddc4-43f9-a781-f72019388548,1992-06-30,NaN,999-27-3385,S99971451,X53218815X,Mr.,Jayson808,Fadel536,NaN,...,Springfield Massachusetts US,1056 Harris Lane Suite 70,Chicopee,Massachusetts,Hampden County,1020.0,42.181642,-72.608842,577445.86,3528.84
3,199c586f-af16-4091-9998-ee4cfc02ee7a,2004-01-09,NaN,999-73-2461,S99956432,NaN,NaN,Jimmie93,Harris789,NaN,...,Worcester Massachusetts US,201 Mitchell Lodge Unit 67,Pembroke,Massachusetts,Plymouth County,NaN,42.075292,-70.757035,336701.72,2705.64
4,353016ea-a0ff-4154-85bb-1cf8b6cedf20,1996-11-15,NaN,999-60-7372,S99917327,X58903159X,Mr.,Gregorio366,Auer97,NaN,...,Patras Achaea GR,1050 Lindgren Extension Apt 38,Boston,Massachusetts,Suffolk County,2135.0,42.352434,-71.028610,484076.34,3043.04
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12347,eec72396-d10e-4b8e-8fb6-25fc1e396031,1962-08-17,NaN,999-52-5112,S99981543,X65879689X,Mrs.,Althea11,O'Hara248,NaN,...,Delhi Delhi IN,682 Koss Trafficway Apt 65,Wellesley,Massachusetts,Norfolk County,2457.0,42.294175,-71.259364,1332256.08,7157.90
12348,3daf2b90-c439-469f-bb44-fb6109c6a8a7,1918-10-20,1999-11-01,999-25-1153,S99951005,X1374154X,Mrs.,Tarah156,Shields502,NaN,...,New Bedford Massachusetts US,308 Huels Grove Apt 18,Waltham,Massachusetts,Middlesex County,2453.0,42.372663,-71.209053,1557234.62,27214.48
12349,cb03d88e-7f08-46a1-be78-70d9fe8262a7,1918-10-20,2009-08-07,999-41-7538,S99938113,X29702266X,Mrs.,Penny812,Pacocha935,NaN,...,Marshfield Massachusetts US,349 Breitenberg Walk Suite 26,Waltham,Massachusetts,Middlesex County,2452.0,42.396186,-71.217928,1347930.08,26001.41
12350,435be43f-cb57-442f-aeb2-51018b030ed0,1918-10-20,NaN,999-26-4571,S99995011,X51104300X,Ms.,Cherlyn665,Quitzon246,NaN,...,Holyoke Massachusetts US,237 Miller Avenue,Waltham,Massachusetts,Middlesex County,2452.0,42.366768,-71.196715,1585631.29,32006.97


In [168]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 12352 entries, 0 to 12351
Data columns (total 25 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Id                   12352 non-null  str    
 1   BIRTHDATE            12352 non-null  str    
 2   DEATHDATE            2352 non-null   str    
 3   SSN                  12352 non-null  str    
 4   DRIVERS              10399 non-null  str    
 5   PASSPORT             9845 non-null   str    
 6   PREFIX               10110 non-null  str    
 7   FIRST                12352 non-null  str    
 8   LAST                 12352 non-null  str    
 9   SUFFIX               124 non-null    str    
 10  MAIDEN               3540 non-null   str    
 11  MARITAL              8833 non-null   str    
 12  RACE                 12352 non-null  str    
 13  ETHNICITY            12352 non-null  str    
 14  GENDER               12352 non-null  str    
 15  BIRTHPLACE           12352 non-null  str    
 1

In [169]:
df.describe(include='all')

,Id,BIRTHDATE,DEATHDATE,SSN,DRIVERS,PASSPORT,PREFIX,FIRST,LAST,SUFFIX,...,BIRTHPLACE,ADDRESS,CITY,STATE,COUNTY,ZIP,LAT,LON,HEALTHCARE_EXPENSES,HEALTHCARE_COVERAGE
count,12352,12352,2352,12352,10399,9845,10110,12352,12352,124,...,12352,12352,12352,12352,12352,6700.000000,12352.000000,12352.000000,1.235200e+04,1.235200e+04
unique,12352,8571,1928,12254,9847,9844,3,4220,887,3,...,427,12352,331,1,14,NaN,NaN,NaN,NaN,NaN
top,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,1931-04-13,2020-03-15,999-27-3677,S99911355,X32013195X,Mr.,Colby655,Rau926,JD,...,Boston Massachusetts US,888 Hickle Ferry Suite 38,Boston,Massachusetts,Middlesex County,NaN,NaN,NaN,NaN,NaN
freq,1,20,20,3,3,2,5008,15,39,46,...,1103,1,1190,12352,2812,NaN,NaN,NaN,NaN,NaN
mean,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,1991.142985,42.269817,-71.361474,8.233402e+05,1.314389e+04
std,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,446.364790,0.282758,0.605881,5.609016e+05,4.392334e+04
min,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,1001.000000,41.244087,-73.478405,5.000000e+02,0.000000e+00
25%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,1757.000000,42.122693,-71.563264,2.712708e+05,3.228080e+03
50%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,2116.000000,42.308687,-71.158615,8.224834e+05,6.138940e+03
75%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,2199.000000,42.449323,-71.025688,1.359019e+06,1.144249e+04


In [170]:
df['BIRTHDATE'] = pd.to_datetime(df['BIRTHDATE'], errors='coerce')
df['DEATHDATE'] = pd.to_datetime(df['DEATHDATE'], errors='coerce')

In [171]:
df.head()

,Id,BIRTHDATE,DEATHDATE,SSN,DRIVERS,PASSPORT,PREFIX,FIRST,LAST,SUFFIX,...,BIRTHPLACE,ADDRESS,CITY,STATE,COUNTY,ZIP,LAT,LON,HEALTHCARE_EXPENSES,HEALTHCARE_COVERAGE
0,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,2017-08-24,NaT,999-68-6630,NaN,NaN,NaN,Jacinto644,Kris249,NaN,...,Beverly Massachusetts US,888 Hickle Ferry Suite 38,Springfield,Massachusetts,Hampden County,1106.0,42.151961,-72.598959,8446.49,1499.08
1,067318a4-db8f-447f-8b6e-f2f61e9baaa5,2016-08-01,NaT,999-15-5895,NaN,NaN,NaN,Alva958,Krajcik437,NaN,...,Boston Massachusetts US,1048 Skiles Trailer,Walpole,Massachusetts,Norfolk County,2081.0,42.177370,-71.281353,89893.40,1845.72
2,ae9efba3-ddc4-43f9-a781-f72019388548,1992-06-30,NaT,999-27-3385,S99971451,X53218815X,Mr.,Jayson808,Fadel536,NaN,...,Springfield Massachusetts US,1056 Harris Lane Suite 70,Chicopee,Massachusetts,Hampden County,1020.0,42.181642,-72.608842,577445.86,3528.84
3,199c586f-af16-4091-9998-ee4cfc02ee7a,2004-01-09,NaT,999-73-2461,S99956432,NaN,NaN,Jimmie93,Harris789,NaN,...,Worcester Massachusetts US,201 Mitchell Lodge Unit 67,Pembroke,Massachusetts,Plymouth County,NaN,42.075292,-70.757035,336701.72,2705.64
4,353016ea-a0ff-4154-85bb-1cf8b6cedf20,1996-11-15,NaT,999-60-7372,S99917327,X58903159X,Mr.,Gregorio366,Auer97,NaN,...,Patras Achaea GR,1050 Lindgren Extension Apt 38,Boston,Massachusetts,Suffolk County,2135.0,42.352434,-71.028610,484076.34,3043.04


In [172]:
df.columns

Index(['Id', 'BIRTHDATE', 'DEATHDATE', 'SSN', 'DRIVERS', 'PASSPORT', 'PREFIX',
       'FIRST', 'LAST', 'SUFFIX', 'MAIDEN', 'MARITAL', 'RACE', 'ETHNICITY',
       'GENDER', 'BIRTHPLACE', 'ADDRESS', 'CITY', 'STATE', 'COUNTY', 'ZIP',
       'LAT', 'LON', 'HEALTHCARE_EXPENSES', 'HEALTHCARE_COVERAGE'],
      dtype='str')

In [173]:
df=df.copy()

In [174]:
df['age'] = (pd.to_datetime('today') - df['BIRTHDATE']).dt.days // 365

In [175]:
df

,Id,BIRTHDATE,DEATHDATE,SSN,DRIVERS,PASSPORT,PREFIX,FIRST,LAST,SUFFIX,...,ADDRESS,CITY,STATE,COUNTY,ZIP,LAT,LON,HEALTHCARE_EXPENSES,HEALTHCARE_COVERAGE,age
0,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,2017-08-24,NaT,999-68-6630,NaN,NaN,NaN,Jacinto644,Kris249,NaN,...,888 Hickle Ferry Suite 38,Springfield,Massachusetts,Hampden County,1106.0,42.151961,-72.598959,8446.49,1499.08,8
1,067318a4-db8f-447f-8b6e-f2f61e9baaa5,2016-08-01,NaT,999-15-5895,NaN,NaN,NaN,Alva958,Krajcik437,NaN,...,1048 Skiles Trailer,Walpole,Massachusetts,Norfolk County,2081.0,42.177370,-71.281353,89893.40,1845.72,10
2,ae9efba3-ddc4-43f9-a781-f72019388548,1992-06-30,NaT,999-27-3385,S99971451,X53218815X,Mr.,Jayson808,Fadel536,NaN,...,1056 Harris Lane Suite 70,Chicopee,Massachusetts,Hampden County,1020.0,42.181642,-72.608842,577445.86,3528.84,34
3,199c586f-af16-4091-9998-ee4cfc02ee7a,2004-01-09,NaT,999-73-2461,S99956432,NaN,NaN,Jimmie93,Harris789,NaN,...,201 Mitchell Lodge Unit 67,Pembroke,Massachusetts,Plymouth County,NaN,42.075292,-70.757035,336701.72,2705.64,22
4,353016ea-a0ff-4154-85bb-1cf8b6cedf20,1996-11-15,NaT,999-60-7372,S99917327,X58903159X,Mr.,Gregorio366,Auer97,NaN,...,1050 Lindgren Extension Apt 38,Boston,Massachusetts,Suffolk County,2135.0,42.352434,-71.028610,484076.34,3043.04,29
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12347,eec72396-d10e-4b8e-8fb6-25fc1e396031,1962-08-17,NaT,999-52-5112,S99981543,X65879689X,Mrs.,Althea11,O'Hara248,NaN,...,682 Koss Trafficway Apt 65,Wellesley,Massachusetts,Norfolk County,2457.0,42.294175,-71.259364,1332256.08,7157.90,64
12348,3daf2b90-c439-469f-bb44-fb6109c6a8a7,1918-10-20,1999-11-01,999-25-1153,S99951005,X1374154X,Mrs.,Tarah156,Shields502,NaN,...,308 Huels Grove Apt 18,Waltham,Massachusetts,Middlesex County,2453.0,42.372663,-71.209053,1557234.62,27214.48,107
12349,cb03d88e-7f08-46a1-be78-70d9fe8262a7,1918-10-20,2009-08-07,999-41-7538,S99938113,X29702266X,Mrs.,Penny812,Pacocha935,NaN,...,349 Breitenberg Walk Suite 26,Waltham,Massachusetts,Middlesex County,2452.0,42.396186,-71.217928,1347930.08,26001.41,107
12350,435be43f-cb57-442f-aeb2-51018b030ed0,1918-10-20,NaT,999-26-4571,S99995011,X51104300X,Ms.,Cherlyn665,Quitzon246,NaN,...,237 Miller Avenue,Waltham,Massachusetts,Middlesex County,2452.0,42.366768,-71.196715,1585631.29,32006.97,107


In [176]:
df['age'].describe()

count    12352.000000
mean        54.029226
std         28.032928
min          6.000000
25%         31.000000
50%         54.000000
75%         73.000000
max        116.000000
Name: age, dtype: float64

In [177]:
df['DEATHDATE'].isna().value_counts()

DEATHDATE
True     10000
False     2352
Name: count, dtype: int64

In [178]:
####fil

In [179]:

df['is_alive'] = df['DEATHDATE'].isna()

In [180]:
df

,Id,BIRTHDATE,DEATHDATE,SSN,DRIVERS,PASSPORT,PREFIX,FIRST,LAST,SUFFIX,...,CITY,STATE,COUNTY,ZIP,LAT,LON,HEALTHCARE_EXPENSES,HEALTHCARE_COVERAGE,age,is_alive
0,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,2017-08-24,NaT,999-68-6630,NaN,NaN,NaN,Jacinto644,Kris249,NaN,...,Springfield,Massachusetts,Hampden County,1106.0,42.151961,-72.598959,8446.49,1499.08,8,True
1,067318a4-db8f-447f-8b6e-f2f61e9baaa5,2016-08-01,NaT,999-15-5895,NaN,NaN,NaN,Alva958,Krajcik437,NaN,...,Walpole,Massachusetts,Norfolk County,2081.0,42.177370,-71.281353,89893.40,1845.72,10,True
2,ae9efba3-ddc4-43f9-a781-f72019388548,1992-06-30,NaT,999-27-3385,S99971451,X53218815X,Mr.,Jayson808,Fadel536,NaN,...,Chicopee,Massachusetts,Hampden County,1020.0,42.181642,-72.608842,577445.86,3528.84,34,True
3,199c586f-af16-4091-9998-ee4cfc02ee7a,2004-01-09,NaT,999-73-2461,S99956432,NaN,NaN,Jimmie93,Harris789,NaN,...,Pembroke,Massachusetts,Plymouth County,NaN,42.075292,-70.757035,336701.72,2705.64,22,True
4,353016ea-a0ff-4154-85bb-1cf8b6cedf20,1996-11-15,NaT,999-60-7372,S99917327,X58903159X,Mr.,Gregorio366,Auer97,NaN,...,Boston,Massachusetts,Suffolk County,2135.0,42.352434,-71.028610,484076.34,3043.04,29,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12347,eec72396-d10e-4b8e-8fb6-25fc1e396031,1962-08-17,NaT,999-52-5112,S99981543,X65879689X,Mrs.,Althea11,O'Hara248,NaN,...,Wellesley,Massachusetts,Norfolk County,2457.0,42.294175,-71.259364,1332256.08,7157.90,64,True
12348,3daf2b90-c439-469f-bb44-fb6109c6a8a7,1918-10-20,1999-11-01,999-25-1153,S99951005,X1374154X,Mrs.,Tarah156,Shields502,NaN,...,Waltham,Massachusetts,Middlesex County,2453.0,42.372663,-71.209053,1557234.62,27214.48,107,False
12349,cb03d88e-7f08-46a1-be78-70d9fe8262a7,1918-10-20,2009-08-07,999-41-7538,S99938113,X29702266X,Mrs.,Penny812,Pacocha935,NaN,...,Waltham,Massachusetts,Middlesex County,2452.0,42.396186,-71.217928,1347930.08,26001.41,107,False
12350,435be43f-cb57-442f-aeb2-51018b030ed0,1918-10-20,NaT,999-26-4571,S99995011,X51104300X,Ms.,Cherlyn665,Quitzon246,NaN,...,Waltham,Massachusetts,Middlesex County,2452.0,42.366768,-71.196715,1585631.29,32006.97,107,True


In [181]:
df['GENDER'].value_counts()

GENDER
F    6253
M    6099
Name: count, dtype: int64

In [182]:
df.columns

Index(['Id', 'BIRTHDATE', 'DEATHDATE', 'SSN', 'DRIVERS', 'PASSPORT', 'PREFIX',
       'FIRST', 'LAST', 'SUFFIX', 'MAIDEN', 'MARITAL', 'RACE', 'ETHNICITY',
       'GENDER', 'BIRTHPLACE', 'ADDRESS', 'CITY', 'STATE', 'COUNTY', 'ZIP',
       'LAT', 'LON', 'HEALTHCARE_EXPENSES', 'HEALTHCARE_COVERAGE', 'age',
       'is_alive'],
      dtype='str')

In [183]:
columns_to_keep=['Id', 'BIRTHDATE', 'DEATHDATE','FIRST', 'LAST','MARITAL', 'RACE', 'ETHNICITY',
       'GENDER', 'BIRTHPLACE', 'ADDRESS', 'CITY', 'STATE', 'COUNTY','HEALTHCARE_EXPENSES', 'HEALTHCARE_COVERAGE', 'age',
       'is_alive']

In [184]:
columns_to_dlt=['SSN', 'DRIVERS', 'PASSPORT', 'PREFIX','SUFFIX', 'MAIDEN','ZIP',
       'LAT', 'LON',]

In [185]:
df_modified=df.drop(columns=columns_to_dlt,errors='ignore')

In [186]:

df_modified

,Id,BIRTHDATE,DEATHDATE,FIRST,LAST,MARITAL,RACE,ETHNICITY,GENDER,BIRTHPLACE,ADDRESS,CITY,STATE,COUNTY,HEALTHCARE_EXPENSES,HEALTHCARE_COVERAGE,age,is_alive
0,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,2017-08-24,NaT,Jacinto644,Kris249,NaN,white,nonhispanic,M,Beverly Massachusetts US,888 Hickle Ferry Suite 38,Springfield,Massachusetts,Hampden County,8446.49,1499.08,8,True
1,067318a4-db8f-447f-8b6e-f2f61e9baaa5,2016-08-01,NaT,Alva958,Krajcik437,NaN,white,nonhispanic,F,Boston Massachusetts US,1048 Skiles Trailer,Walpole,Massachusetts,Norfolk County,89893.40,1845.72,10,True
2,ae9efba3-ddc4-43f9-a781-f72019388548,1992-06-30,NaT,Jayson808,Fadel536,S,white,nonhispanic,M,Springfield Massachusetts US,1056 Harris Lane Suite 70,Chicopee,Massachusetts,Hampden County,577445.86,3528.84,34,True
3,199c586f-af16-4091-9998-ee4cfc02ee7a,2004-01-09,NaT,Jimmie93,Harris789,NaN,white,nonhispanic,F,Worcester Massachusetts US,201 Mitchell Lodge Unit 67,Pembroke,Massachusetts,Plymouth County,336701.72,2705.64,22,True
4,353016ea-a0ff-4154-85bb-1cf8b6cedf20,1996-11-15,NaT,Gregorio366,Auer97,NaN,white,nonhispanic,M,Patras Achaea GR,1050 Lindgren Extension Apt 38,Boston,Massachusetts,Suffolk County,484076.34,3043.04,29,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12347,eec72396-d10e-4b8e-8fb6-25fc1e396031,1962-08-17,NaT,Althea11,O'Hara248,M,asian,nonhispanic,F,Delhi Delhi IN,682 Koss Trafficway Apt 65,Wellesley,Massachusetts,Norfolk County,1332256.08,7157.90,64,True
12348,3daf2b90-c439-469f-bb44-fb6109c6a8a7,1918-10-20,1999-11-01,Tarah156,Shields502,M,white,nonhispanic,F,New Bedford Massachusetts US,308 Huels Grove Apt 18,Waltham,Massachusetts,Middlesex County,1557234.62,27214.48,107,False
12349,cb03d88e-7f08-46a1-be78-70d9fe8262a7,1918-10-20,2009-08-07,Penny812,Pacocha935,M,white,nonhispanic,F,Marshfield Massachusetts US,349 Breitenberg Walk Suite 26,Waltham,Massachusetts,Middlesex County,1347930.08,26001.41,107,False
12350,435be43f-cb57-442f-aeb2-51018b030ed0,1918-10-20,NaT,Cherlyn665,Quitzon246,S,white,nonhispanic,F,Holyoke Massachusetts US,237 Miller Avenue,Waltham,Massachusetts,Middlesex County,1585631.29,32006.97,107,True


In [187]:

df_modified.info()

<class 'pandas.DataFrame'>
RangeIndex: 12352 entries, 0 to 12351
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Id                   12352 non-null  str           
 1   BIRTHDATE            12352 non-null  datetime64[us]
 2   DEATHDATE            2352 non-null   datetime64[us]
 3   FIRST                12352 non-null  str           
 4   LAST                 12352 non-null  str           
 5   MARITAL              8833 non-null   str           
 6   RACE                 12352 non-null  str           
 7   ETHNICITY            12352 non-null  str           
 8   GENDER               12352 non-null  str           
 9   BIRTHPLACE           12352 non-null  str           
 10  ADDRESS              12352 non-null  str           
 11  CITY                 12352 non-null  str           
 12  STATE                12352 non-null  str           
 13  COUNTY               12352 non-null  str  

In [188]:
df_modified['DEATHDATE'].isna().value_counts()

DEATHDATE
True     10000
False     2352
Name: count, dtype: int64

In [189]:
df_modified.groupby('GENDER').size()

GENDER
F    6253
M    6099
dtype: int64

In [190]:
df_modified['GENDER'] = df_modified['GENDER'].replace({
    'F': 'FEMALE',
    'M': 'MALE'
})

In [191]:
df_modified.head()

,Id,BIRTHDATE,DEATHDATE,FIRST,LAST,MARITAL,RACE,ETHNICITY,GENDER,BIRTHPLACE,ADDRESS,CITY,STATE,COUNTY,HEALTHCARE_EXPENSES,HEALTHCARE_COVERAGE,age,is_alive
0,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,2017-08-24,NaT,Jacinto644,Kris249,NaN,white,nonhispanic,MALE,Beverly Massachusetts US,888 Hickle Ferry Suite 38,Springfield,Massachusetts,Hampden County,8446.49,1499.08,8,True
1,067318a4-db8f-447f-8b6e-f2f61e9baaa5,2016-08-01,NaT,Alva958,Krajcik437,NaN,white,nonhispanic,FEMALE,Boston Massachusetts US,1048 Skiles Trailer,Walpole,Massachusetts,Norfolk County,89893.40,1845.72,10,True
2,ae9efba3-ddc4-43f9-a781-f72019388548,1992-06-30,NaT,Jayson808,Fadel536,S,white,nonhispanic,MALE,Springfield Massachusetts US,1056 Harris Lane Suite 70,Chicopee,Massachusetts,Hampden County,577445.86,3528.84,34,True
3,199c586f-af16-4091-9998-ee4cfc02ee7a,2004-01-09,NaT,Jimmie93,Harris789,NaN,white,nonhispanic,FEMALE,Worcester Massachusetts US,201 Mitchell Lodge Unit 67,Pembroke,Massachusetts,Plymouth County,336701.72,2705.64,22,True
4,353016ea-a0ff-4154-85bb-1cf8b6cedf20,1996-11-15,NaT,Gregorio366,Auer97,NaN,white,nonhispanic,MALE,Patras Achaea GR,1050 Lindgren Extension Apt 38,Boston,Massachusetts,Suffolk County,484076.34,3043.04,29,True


In [192]:
df_modified.groupby('GENDER').size()

GENDER
FEMALE    6253
MALE      6099
dtype: int64

In [193]:
df_modified.groupby('GENDER')['age'].mean()

GENDER
FEMALE    52.673277
MALE      55.419413
Name: age, dtype: float64

#multilple aggregates

In [194]:
df_modified.groupby('GENDER')['age'].agg(['count', 'mean', 'max', 'min'])

,count,mean,max,min
GENDER,,,,
FEMALE,6253,52.673277,116,6
MALE,6099,55.419413,116,6


In [195]:
df_modified['age_group'] = pd.cut(df_modified['age'], bins=[0,18,40,60,100], labels=['0-18','19-40','41-60','60+'])

In [196]:
df_modified.head(2)

,Id,BIRTHDATE,DEATHDATE,FIRST,LAST,MARITAL,RACE,ETHNICITY,GENDER,BIRTHPLACE,ADDRESS,CITY,STATE,COUNTY,HEALTHCARE_EXPENSES,HEALTHCARE_COVERAGE,age,is_alive,age_group
0,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,2017-08-24,NaT,Jacinto644,Kris249,NaN,white,nonhispanic,MALE,Beverly Massachusetts US,888 Hickle Ferry Suite 38,Springfield,Massachusetts,Hampden County,8446.49,1499.08,8,True,0-18
1,067318a4-db8f-447f-8b6e-f2f61e9baaa5,2016-08-01,NaT,Alva958,Krajcik437,NaN,white,nonhispanic,FEMALE,Boston Massachusetts US,1048 Skiles Trailer,Walpole,Massachusetts,Norfolk County,89893.40,1845.72,10,True,0-18


In [197]:
df_modified['age_group'].value_counts()

age_group
60+      4166
19-40    3028
41-60    2867
0-18     1392
Name: count, dtype: int64

In [198]:
encounters =pd.read_csv("C:\projects\Healthcare-Analytics-Dashboard\python\data\encounters.csv")

In [199]:

encounters

,Id,START,STOP,PATIENT,ORGANIZATION,PROVIDER,PAYER,ENCOUNTERCLASS,CODE,DESCRIPTION,BASE_ENCOUNTER_COST,TOTAL_CLAIM_COST,PAYER_COVERAGE,REASONCODE,REASONDESCRIPTION
0,d5ee30a9-362f-429e-a87a-ee38d999b0a5,2019-02-16T01:02:32Z,2019-02-16T01:17:32Z,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,5103c940-0c08-392f-95cd-446e0cea042a,e2c226c2-3e1e-3d0b-b997-ce9544c10528,7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a,outpatient,185345009,Encounter for symptom,129.16,129.16,69.16,65363002.0,Otitis media
1,6a74fdef-2287-44bf-b9e7-18012376faca,2019-08-02T01:02:32Z,2019-08-02T01:32:32Z,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,0b9f3f7c-8ab6-30a5-b3ae-4dc0e0c00cb3,87c33fc5-3fd1-3c52-815a-b89a1623bb3a,7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a,wellness,410620009,Well child visit (procedure),129.16,129.16,129.16,NaN,NaN
2,8bca6d8a-ab80-4cbf-8abb-46654235f227,2019-10-31T01:02:32Z,2019-10-31T01:17:32Z,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,5103c940-0c08-392f-95cd-446e0cea042a,e2c226c2-3e1e-3d0b-b997-ce9544c10528,7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a,outpatient,185345009,Encounter for symptom,129.16,129.16,69.16,65363002.0,Otitis media
3,821e57ac-9304-46a9-9f9b-83daf60e9e43,2020-01-31T01:02:32Z,2020-01-31T01:17:32Z,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,0b9f3f7c-8ab6-30a5-b3ae-4dc0e0c00cb3,87c33fc5-3fd1-3c52-815a-b89a1623bb3a,7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a,wellness,410620009,Well child visit (procedure),129.16,129.16,129.16,NaN,NaN
4,681c380b-3c84-4c55-80a6-db3d9ea12fee,2020-03-02T01:02:32Z,2020-03-02T01:58:32Z,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,fd328395-ab1d-35c6-a2d0-d05a9a79cf11,9c875a09-93e0-39aa-9260-ad264bbdd3fe,7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a,ambulatory,185345009,Encounter for symptom (procedure),129.16,129.16,69.16,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
321523,9af81365-e947-43b6-b4af-b94efb9182c5,2020-05-19T06:03:58Z,2020-05-19T08:18:58Z,2712205f-755e-4897-acb3-926895b7d635,d5117822-5756-389d-9547-891a372d580f,5ed53ed3-568e-3f70-9797-49ebdc4a74e9,7caa7254-5050-3b5e-9eae-bd5ea30e809c,ambulatory,185347001,Encounter for problem (procedure),129.16,129.16,89.16,NaN,NaN
321524,823b575f-79ce-4946-917a-f7989c0de848,2020-05-26T06:03:58Z,2020-05-26T06:18:58Z,2712205f-755e-4897-acb3-926895b7d635,b60f8daf-0362-3c28-9b54-216e7f3e80bd,23869b55-8805-3f41-8484-3d5c31185f4d,7caa7254-5050-3b5e-9eae-bd5ea30e809c,urgentcare,702927004,Urgent care clinic (procedure),129.16,129.16,0.00,NaN,NaN
321525,d4e1c9e6-2b5e-43a5-9fad-cef200c9e5c5,2020-05-22T06:03:58Z,2020-05-22T09:18:58Z,2712205f-755e-4897-acb3-926895b7d635,d5117822-5756-389d-9547-891a372d580f,5ed53ed3-568e-3f70-9797-49ebdc4a74e9,7caa7254-5050-3b5e-9eae-bd5ea30e809c,ambulatory,185347001,Encounter for problem (procedure),129.16,129.16,89.16,NaN,NaN
321526,f69323ee-fc33-45ff-94ba-914acde61601,2020-05-22T06:03:58Z,2020-05-22T06:18:58Z,2712205f-755e-4897-acb3-926895b7d635,d5117822-5756-389d-9547-891a372d580f,5ed53ed3-568e-3f70-9797-49ebdc4a74e9,7caa7254-5050-3b5e-9eae-bd5ea30e809c,ambulatory,185347001,Encounter for problem (procedure),129.16,129.16,89.16,NaN,NaN


In [200]:
encounters.shape

(321528, 15)

In [201]:
encounters.info()

<class 'pandas.DataFrame'>
RangeIndex: 321528 entries, 0 to 321527
Data columns (total 15 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   Id                   321528 non-null  str    
 1   START                321528 non-null  str    
 2   STOP                 321528 non-null  str    
 3   PATIENT              321528 non-null  str    
 4   ORGANIZATION         321528 non-null  str    
 5   PROVIDER             321528 non-null  str    
 6   PAYER                321528 non-null  str    
 7   ENCOUNTERCLASS       321528 non-null  str    
 8   CODE                 321528 non-null  int64  
 9   DESCRIPTION          321528 non-null  str    
 10  BASE_ENCOUNTER_COST  321528 non-null  float64
 11  TOTAL_CLAIM_COST     321528 non-null  float64
 12  PAYER_COVERAGE       321528 non-null  float64
 13  REASONCODE           71252 non-null   float64
 14  REASONDESCRIPTION    71252 non-null   str    
dtypes: float64(4), int64(1), str

In [202]:
encounters.columns

Index(['Id', 'START', 'STOP', 'PATIENT', 'ORGANIZATION', 'PROVIDER', 'PAYER',
       'ENCOUNTERCLASS', 'CODE', 'DESCRIPTION', 'BASE_ENCOUNTER_COST',
       'TOTAL_CLAIM_COST', 'PAYER_COVERAGE', 'REASONCODE',
       'REASONDESCRIPTION'],
      dtype='str')

In [203]:
df_modified.columns

Index(['Id', 'BIRTHDATE', 'DEATHDATE', 'FIRST', 'LAST', 'MARITAL', 'RACE',
       'ETHNICITY', 'GENDER', 'BIRTHPLACE', 'ADDRESS', 'CITY', 'STATE',
       'COUNTY', 'HEALTHCARE_EXPENSES', 'HEALTHCARE_COVERAGE', 'age',
       'is_alive', 'age_group'],
      dtype='str')

In [204]:
df_modified['Id'].nunique()

12352

In [205]:
encounters['Id'].nunique()

321528

In [206]:
encounters['PATIENT'].isin(df_modified['Id']).value_counts()

PATIENT
True    321528
Name: count, dtype: int64

In [207]:
patient_encounters=df_modified.merge(encounters,left_on='Id',right_on='PATIENT',how='left')

In [208]:
patient_encounters

,Id_x,BIRTHDATE,DEATHDATE,FIRST,LAST,MARITAL,RACE,ETHNICITY,GENDER,BIRTHPLACE,...,PROVIDER,PAYER,ENCOUNTERCLASS,CODE,DESCRIPTION,BASE_ENCOUNTER_COST,TOTAL_CLAIM_COST,PAYER_COVERAGE,REASONCODE,REASONDESCRIPTION
0,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,2017-08-24,NaT,Jacinto644,Kris249,NaN,white,nonhispanic,MALE,Beverly Massachusetts US,...,e2c226c2-3e1e-3d0b-b997-ce9544c10528,7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a,outpatient,185345009.0,Encounter for symptom,129.16,129.16,69.16,65363002.0,Otitis media
1,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,2017-08-24,NaT,Jacinto644,Kris249,NaN,white,nonhispanic,MALE,Beverly Massachusetts US,...,87c33fc5-3fd1-3c52-815a-b89a1623bb3a,7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a,wellness,410620009.0,Well child visit (procedure),129.16,129.16,129.16,NaN,NaN
2,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,2017-08-24,NaT,Jacinto644,Kris249,NaN,white,nonhispanic,MALE,Beverly Massachusetts US,...,e2c226c2-3e1e-3d0b-b997-ce9544c10528,7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a,outpatient,185345009.0,Encounter for symptom,129.16,129.16,69.16,65363002.0,Otitis media
3,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,2017-08-24,NaT,Jacinto644,Kris249,NaN,white,nonhispanic,MALE,Beverly Massachusetts US,...,87c33fc5-3fd1-3c52-815a-b89a1623bb3a,7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a,wellness,410620009.0,Well child visit (procedure),129.16,129.16,129.16,NaN,NaN
4,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,2017-08-24,NaT,Jacinto644,Kris249,NaN,white,nonhispanic,MALE,Beverly Massachusetts US,...,9c875a09-93e0-39aa-9260-ad264bbdd3fe,7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a,ambulatory,185345009.0,Encounter for symptom (procedure),129.16,129.16,69.16,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
321545,2712205f-755e-4897-acb3-926895b7d635,1941-02-11,NaT,Emory494,Turcotte120,S,white,nonhispanic,MALE,Medford Massachusetts US,...,5ed53ed3-568e-3f70-9797-49ebdc4a74e9,7caa7254-5050-3b5e-9eae-bd5ea30e809c,ambulatory,185347001.0,Encounter for problem (procedure),129.16,129.16,89.16,NaN,NaN
321546,2712205f-755e-4897-acb3-926895b7d635,1941-02-11,NaT,Emory494,Turcotte120,S,white,nonhispanic,MALE,Medford Massachusetts US,...,23869b55-8805-3f41-8484-3d5c31185f4d,7caa7254-5050-3b5e-9eae-bd5ea30e809c,urgentcare,702927004.0,Urgent care clinic (procedure),129.16,129.16,0.00,NaN,NaN
321547,2712205f-755e-4897-acb3-926895b7d635,1941-02-11,NaT,Emory494,Turcotte120,S,white,nonhispanic,MALE,Medford Massachusetts US,...,5ed53ed3-568e-3f70-9797-49ebdc4a74e9,7caa7254-5050-3b5e-9eae-bd5ea30e809c,ambulatory,185347001.0,Encounter for problem (procedure),129.16,129.16,89.16,NaN,NaN
321548,2712205f-755e-4897-acb3-926895b7d635,1941-02-11,NaT,Emory494,Turcotte120,S,white,nonhispanic,MALE,Medford Massachusetts US,...,5ed53ed3-568e-3f70-9797-49ebdc4a74e9,7caa7254-5050-3b5e-9eae-bd5ea30e809c,ambulatory,185347001.0,Encounter for problem (procedure),129.16,129.16,89.16,NaN,NaN


In [209]:
patient_encounters.columns

Index(['Id_x', 'BIRTHDATE', 'DEATHDATE', 'FIRST', 'LAST', 'MARITAL', 'RACE',
       'ETHNICITY', 'GENDER', 'BIRTHPLACE', 'ADDRESS', 'CITY', 'STATE',
       'COUNTY', 'HEALTHCARE_EXPENSES', 'HEALTHCARE_COVERAGE', 'age',
       'is_alive', 'age_group', 'Id_y', 'START', 'STOP', 'PATIENT',
       'ORGANIZATION', 'PROVIDER', 'PAYER', 'ENCOUNTERCLASS', 'CODE',
       'DESCRIPTION', 'BASE_ENCOUNTER_COST', 'TOTAL_CLAIM_COST',
       'PAYER_COVERAGE', 'REASONCODE', 'REASONDESCRIPTION'],
      dtype='str')

In [210]:
patient_encounters.shape

(321550, 34)

In [211]:
encounter_counts = (
    patient_encounters
    .groupby('Id_x')
    .size()
)

In [212]:
encounter_counts

Id_x
0000b247-1def-417a-a783-41c8682be022     3
00049ee8-5953-4edd-a277-b9c1b1a7f16b     8
000769a6-23a7-426e-a264-cb0e509b2da2     4
00079a57-24a8-430f-b4f8-a1cf34f90060    20
0008a63c-c95c-46c2-9ef3-831d68892019     5
                                        ..
ffd3d544-1fcd-4a87-9514-fa6c37409cbc    75
ffd86fda-ebb9-400e-9fe3-ea1a1037dbad    14
ffdbbb1b-745e-4e38-ade2-a19d6e778fee    34
ffdf0900-bc4b-4f81-b95b-1ea57da21e07    36
ffea1d41-7562-499d-af1b-284ac72a7c3c     5
Length: 12352, dtype: int64

In [213]:
encounter_counts.head()

Id_x
0000b247-1def-417a-a783-41c8682be022     3
00049ee8-5953-4edd-a277-b9c1b1a7f16b     8
000769a6-23a7-426e-a264-cb0e509b2da2     4
00079a57-24a8-430f-b4f8-a1cf34f90060    20
0008a63c-c95c-46c2-9ef3-831d68892019     5
dtype: int64

In [214]:
visit_frequency = (
    patient_encounters
    .groupby('Id_x')
    .size()
    .reset_index(name='visit_count')
)

In [215]:
visit_frequency

,Id_x,visit_count
0,0000b247-1def-417a-a783-41c8682be022,3
1,00049ee8-5953-4edd-a277-b9c1b1a7f16b,8
2,000769a6-23a7-426e-a264-cb0e509b2da2,4
3,00079a57-24a8-430f-b4f8-a1cf34f90060,20
4,0008a63c-c95c-46c2-9ef3-831d68892019,5
...,...,...
12347,ffd3d544-1fcd-4a87-9514-fa6c37409cbc,75
12348,ffd86fda-ebb9-400e-9fe3-ea1a1037dbad,14
12349,ffdbbb1b-745e-4e38-ade2-a19d6e778fee,34
12350,ffdf0900-bc4b-4f81-b95b-1ea57da21e07,36


Who are the highest-utilization patients?

In [216]:
visit_frequency.sort_values('visit_count',ascending=False).head(10)

,Id_x,visit_count
12310,fef74b94-c2d9-4e22-bc4f-382c82152126,825
4993,66f888db-3cd2-463e-8fa5-43c50dec2a3c,658
8817,b7e10f32-477b-4a19-b6e4-5dca83d86574,611
3779,4d9a197a-0469-4144-9030-0c04654cb4a8,605
4156,55786035-203e-463f-87cf-3fc6cbf8b119,584
7499,9cd159bb-f711-45b3-a353-1e354587b488,580
12035,f93c04a1-b7b2-420d-9545-feab6d7dd1f0,544
10925,e396584a-3b3a-44e2-838f-be05097ef6bb,535
1845,2606458d-2648-49aa-af43-594a26be575f,525
10456,d988bba1-094c-4bfa-a74f-befd2ad8c60b,519


In [217]:
patient_encounters[['Id_x', 'Id_y', 'PATIENT']].head()

,Id_x,Id_y,PATIENT
0,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,d5ee30a9-362f-429e-a87a-ee38d999b0a5,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271
1,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,6a74fdef-2287-44bf-b9e7-18012376faca,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271
2,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,8bca6d8a-ab80-4cbf-8abb-46654235f227,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271
3,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,821e57ac-9304-46a9-9f9b-83daf60e9e43,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271
4,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,681c380b-3c84-4c55-80a6-db3d9ea12fee,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271


In [218]:
visit_frequency = visit_frequency.rename(
    columns={'Id_x': 'patient_id',
            'Id_y':'encounter_id'}
)

In [219]:
visit_frequency

,patient_id,visit_count
0,0000b247-1def-417a-a783-41c8682be022,3
1,00049ee8-5953-4edd-a277-b9c1b1a7f16b,8
2,000769a6-23a7-426e-a264-cb0e509b2da2,4
3,00079a57-24a8-430f-b4f8-a1cf34f90060,20
4,0008a63c-c95c-46c2-9ef3-831d68892019,5
...,...,...
12347,ffd3d544-1fcd-4a87-9514-fa6c37409cbc,75
12348,ffd86fda-ebb9-400e-9fe3-ea1a1037dbad,14
12349,ffdbbb1b-745e-4e38-ade2-a19d6e778fee,34
12350,ffdf0900-bc4b-4f81-b95b-1ea57da21e07,36


In [220]:
encounters['ENCOUNTERCLASS'].value_counts()

ENCOUNTERCLASS
wellness      156219
ambulatory     86069
outpatient     58367
inpatient       9584
emergency       7233
urgentcare      4056
Name: count, dtype: int64

In [221]:
#"How many encounters of each type does each patient have?"

In [222]:
encounters.columns

Index(['Id', 'START', 'STOP', 'PATIENT', 'ORGANIZATION', 'PROVIDER', 'PAYER',
       'ENCOUNTERCLASS', 'CODE', 'DESCRIPTION', 'BASE_ENCOUNTER_COST',
       'TOTAL_CLAIM_COST', 'PAYER_COVERAGE', 'REASONCODE',
       'REASONDESCRIPTION'],
      dtype='str')

In [223]:
encounter_by_type=(encounters.groupby(['PATIENT','ENCOUNTERCLASS']).size().reset_index(name='encounter_count'))

In [224]:
encounter_by_type

,PATIENT,ENCOUNTERCLASS,encounter_count
0,0000b247-1def-417a-a783-41c8682be022,ambulatory,2
1,0000b247-1def-417a-a783-41c8682be022,wellness,1
2,00049ee8-5953-4edd-a277-b9c1b1a7f16b,ambulatory,5
3,00049ee8-5953-4edd-a277-b9c1b1a7f16b,outpatient,1
4,00049ee8-5953-4edd-a277-b9c1b1a7f16b,wellness,2
...,...,...,...
38873,ffdf0900-bc4b-4f81-b95b-1ea57da21e07,outpatient,1
38874,ffdf0900-bc4b-4f81-b95b-1ea57da21e07,wellness,32
38875,ffea1d41-7562-499d-af1b-284ac72a7c3c,ambulatory,2
38876,ffea1d41-7562-499d-af1b-284ac72a7c3c,emergency,2


In [225]:
encounter_by_type.sort_values(
    'encounter_count',
    ascending=False
).head(10)

,PATIENT,ENCOUNTERCLASS,encounter_count
38733,fef74b94-c2d9-4e22-bc4f-382c82152126,ambulatory,772
15591,66f888db-3cd2-463e-8fa5-43c50dec2a3c,outpatient,503
32838,d988bba1-094c-4bfa-a74f-befd2ad8c60b,outpatient,406
156,00c4a300-1dd2-40cb-a7c5-98d8dbb5e5d0,outpatient,387
8599,39768268-b4a5-48c5-ae71-ac65f07deb30,outpatient,374
35916,edc665ca-8db1-4d11-92ee-3110218766ea,outpatient,368
26614,b1385686-0b35-4366-a4f3-5b8779d18ecc,outpatient,360
8521,38967272-1a82-4d94-ac46-9a813f953280,outpatient,331
15178,643f342f-b55d-47ea-88fe-911d33a7a4d1,outpatient,329
30389,ca073c5b-6c05-4ae9-ac65-d567174191ee,outpatient,325


In [226]:
visit_frequency['visit_count'].describe()

count    12352.000000
mean        26.032222
std         44.923710
min          1.000000
25%          5.000000
50%         10.000000
75%         33.000000
max        825.000000
Name: visit_count, dtype: float64

In [227]:
patient_encounters[['Id_x', 'Id_y', 'PATIENT']].head()

,Id_x,Id_y,PATIENT
0,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,d5ee30a9-362f-429e-a87a-ee38d999b0a5,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271
1,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,6a74fdef-2287-44bf-b9e7-18012376faca,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271
2,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,8bca6d8a-ab80-4cbf-8abb-46654235f227,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271
3,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,821e57ac-9304-46a9-9f9b-83daf60e9e43,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271
4,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,681c380b-3c84-4c55-80a6-db3d9ea12fee,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271


In [228]:
encounters['ENCOUNTERCLASS'].value_counts()

ENCOUNTERCLASS
wellness      156219
ambulatory     86069
outpatient     58367
inpatient       9584
emergency       7233
urgentcare      4056
Name: count, dtype: int64

In [229]:
visit_frequency['visit_count'].describe()

count    12352.000000
mean        26.032222
std         44.923710
min          1.000000
25%          5.000000
50%         10.000000
75%         33.000000
max        825.000000
Name: visit_count, dtype: float64

In [230]:
df_modified

,Id,BIRTHDATE,DEATHDATE,FIRST,LAST,MARITAL,RACE,ETHNICITY,GENDER,BIRTHPLACE,ADDRESS,CITY,STATE,COUNTY,HEALTHCARE_EXPENSES,HEALTHCARE_COVERAGE,age,is_alive,age_group
0,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,2017-08-24,NaT,Jacinto644,Kris249,NaN,white,nonhispanic,MALE,Beverly Massachusetts US,888 Hickle Ferry Suite 38,Springfield,Massachusetts,Hampden County,8446.49,1499.08,8,True,0-18
1,067318a4-db8f-447f-8b6e-f2f61e9baaa5,2016-08-01,NaT,Alva958,Krajcik437,NaN,white,nonhispanic,FEMALE,Boston Massachusetts US,1048 Skiles Trailer,Walpole,Massachusetts,Norfolk County,89893.40,1845.72,10,True,0-18
2,ae9efba3-ddc4-43f9-a781-f72019388548,1992-06-30,NaT,Jayson808,Fadel536,S,white,nonhispanic,MALE,Springfield Massachusetts US,1056 Harris Lane Suite 70,Chicopee,Massachusetts,Hampden County,577445.86,3528.84,34,True,19-40
3,199c586f-af16-4091-9998-ee4cfc02ee7a,2004-01-09,NaT,Jimmie93,Harris789,NaN,white,nonhispanic,FEMALE,Worcester Massachusetts US,201 Mitchell Lodge Unit 67,Pembroke,Massachusetts,Plymouth County,336701.72,2705.64,22,True,19-40
4,353016ea-a0ff-4154-85bb-1cf8b6cedf20,1996-11-15,NaT,Gregorio366,Auer97,NaN,white,nonhispanic,MALE,Patras Achaea GR,1050 Lindgren Extension Apt 38,Boston,Massachusetts,Suffolk County,484076.34,3043.04,29,True,19-40
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12347,eec72396-d10e-4b8e-8fb6-25fc1e396031,1962-08-17,NaT,Althea11,O'Hara248,M,asian,nonhispanic,FEMALE,Delhi Delhi IN,682 Koss Trafficway Apt 65,Wellesley,Massachusetts,Norfolk County,1332256.08,7157.90,64,True,60+
12348,3daf2b90-c439-469f-bb44-fb6109c6a8a7,1918-10-20,1999-11-01,Tarah156,Shields502,M,white,nonhispanic,FEMALE,New Bedford Massachusetts US,308 Huels Grove Apt 18,Waltham,Massachusetts,Middlesex County,1557234.62,27214.48,107,False,NaN
12349,cb03d88e-7f08-46a1-be78-70d9fe8262a7,1918-10-20,2009-08-07,Penny812,Pacocha935,M,white,nonhispanic,FEMALE,Marshfield Massachusetts US,349 Breitenberg Walk Suite 26,Waltham,Massachusetts,Middlesex County,1347930.08,26001.41,107,False,NaN
12350,435be43f-cb57-442f-aeb2-51018b030ed0,1918-10-20,NaT,Cherlyn665,Quitzon246,S,white,nonhispanic,FEMALE,Holyoke Massachusetts US,237 Miller Avenue,Waltham,Massachusetts,Middlesex County,1585631.29,32006.97,107,True,NaN


In [231]:
patient_utilization= visit_frequency.merge(df_modified[['Id','GENDER','age','age_group']],left_on='patient_id',right_on='Id',how='left')

In [232]:
patient_utilization

,patient_id,visit_count,Id,GENDER,age,age_group
0,0000b247-1def-417a-a783-41c8682be022,3,0000b247-1def-417a-a783-41c8682be022,FEMALE,18,0-18
1,00049ee8-5953-4edd-a277-b9c1b1a7f16b,8,00049ee8-5953-4edd-a277-b9c1b1a7f16b,MALE,41,41-60
2,000769a6-23a7-426e-a264-cb0e509b2da2,4,000769a6-23a7-426e-a264-cb0e509b2da2,FEMALE,79,60+
3,00079a57-24a8-430f-b4f8-a1cf34f90060,20,00079a57-24a8-430f-b4f8-a1cf34f90060,FEMALE,35,19-40
4,0008a63c-c95c-46c2-9ef3-831d68892019,5,0008a63c-c95c-46c2-9ef3-831d68892019,MALE,33,19-40
...,...,...,...,...,...,...
12347,ffd3d544-1fcd-4a87-9514-fa6c37409cbc,75,ffd3d544-1fcd-4a87-9514-fa6c37409cbc,MALE,76,60+
12348,ffd86fda-ebb9-400e-9fe3-ea1a1037dbad,14,ffd86fda-ebb9-400e-9fe3-ea1a1037dbad,FEMALE,34,19-40
12349,ffdbbb1b-745e-4e38-ade2-a19d6e778fee,34,ffdbbb1b-745e-4e38-ade2-a19d6e778fee,MALE,67,60+
12350,ffdf0900-bc4b-4f81-b95b-1ea57da21e07,36,ffdf0900-bc4b-4f81-b95b-1ea57da21e07,MALE,56,41-60


In [233]:
#Remove the duplicate ID

In [234]:
patient_utilization=patient_utilization.drop(columns='Id')


In [235]:
patient_utilization

,patient_id,visit_count,GENDER,age,age_group
0,0000b247-1def-417a-a783-41c8682be022,3,FEMALE,18,0-18
1,00049ee8-5953-4edd-a277-b9c1b1a7f16b,8,MALE,41,41-60
2,000769a6-23a7-426e-a264-cb0e509b2da2,4,FEMALE,79,60+
3,00079a57-24a8-430f-b4f8-a1cf34f90060,20,FEMALE,35,19-40
4,0008a63c-c95c-46c2-9ef3-831d68892019,5,MALE,33,19-40
...,...,...,...,...,...
12347,ffd3d544-1fcd-4a87-9514-fa6c37409cbc,75,MALE,76,60+
12348,ffd86fda-ebb9-400e-9fe3-ea1a1037dbad,14,FEMALE,34,19-40
12349,ffdbbb1b-745e-4e38-ade2-a19d6e778fee,34,MALE,67,60+
12350,ffdf0900-bc4b-4f81-b95b-1ea57da21e07,36,MALE,56,41-60


In [236]:
patient_utilization.shape

(12352, 5)

In [237]:
patient_utilization[['visit_count', 'GENDER', 'age', 'age_group']].isna().sum()

visit_count      0
GENDER           0
age              0
age_group      899
dtype: int64

In [238]:
df_modified[df_modified['age_group'].isna()]['age'].describe()

count    899.000000
mean     108.908788
std        4.539915
min      101.000000
25%      104.000000
50%      110.000000
75%      112.000000
max      116.000000
Name: age, dtype: float64

In [239]:
df_modified[df_modified['age_group'].isna()]['age'].value_counts().sort_index()

age
101     66
102     37
103     58
104     65
105     29
106     45
107      8
108     54
109     46
110    106
111     75
112     87
113     90
114     41
115     27
116     65
Name: count, dtype: int64

In [240]:
df_modified['age_group'] = pd.cut(
    df_modified['age'],
    bins=[0, 18, 40, 60, 100, float('inf')],
    labels=['0-18', '19-40', '41-60', '61-100', '100+'],
    include_lowest=True
)

In [241]:
df_modified[df_modified['age_group'].isna()]['age'].describe()

count    0.0
mean     NaN
std      NaN
min      NaN
25%      NaN
50%      NaN
75%      NaN
max      NaN
Name: age, dtype: float64

In [242]:
df_modified['age_group'].isna().sum()

np.int64(0)

In [243]:
df_modified['age_group'].value_counts().sort_index()

age_group
0-18      1392
19-40     3028
41-60     2867
61-100    4166
100+       899
Name: count, dtype: int64

In [244]:
#Does utilization differ by age group?
#Do older patients have more healthcare encounters than younger patients?

In [245]:
"""We need to:

Put patients into age groups.
Take the visit_count for each patient.
Group those patients by age_group.
Calculate summary(average) statistics for each group.

You already learned groupby()."""

'We need to:\n\nPut patients into age groups.\nTake the visit_count for each patient.\nGroup those patients by age_group.\nCalculate summary(average) statistics for each group.\n\nYou already learned groupby().'

In [246]:
patient_utilization.head(2)

,patient_id,visit_count,GENDER,age,age_group
0,0000b247-1def-417a-a783-41c8682be022,3,FEMALE,18,0-18
1,00049ee8-5953-4edd-a277-b9c1b1a7f16b,8,MALE,41,41-60


In [247]:
patient_utilization.groupby('age_group')['visit_count'].mean()

age_group
0-18      4.590517
19-40     9.612946
41-60    19.094175
60+      41.216275
Name: visit_count, dtype: float64

In [248]:
patient_utilization.groupby('age_group')['visit_count'].aggregate(['count','mean','median','max'])

,count,mean,median,max
age_group,,,,
0-18,1392,4.590517,4.0,32
19-40,3028,9.612946,6.0,65
41-60,2867,19.094175,10.0,266
60+,4166,41.216275,23.0,544


In [249]:
age_utilization = (
    patient_utilization
    .groupby('age_group')['visit_count']
    .agg(['count', 'mean', 'median', 'max'])
    .sort_values('mean', ascending=False)
)

In [250]:
age_utilization

,count,mean,median,max
age_group,,,,
60+,4166,41.216275,23.0,544
41-60,2867,19.094175,10.0,266
19-40,3028,9.612946,6.0,65
0-18,1392,4.590517,4.0,32


In [251]:
#Older patients show higher healthcare utilization on average

In [252]:
#Does utilization differ by gender?

In [253]:
patient_utilization.groupby('GENDER')['visit_count'].mean()

GENDER
FEMALE    24.371662
MALE      27.734711
Name: visit_count, dtype: float64

In [254]:
gender_utilization=patient_utilization.groupby('GENDER')['visit_count'].aggregate(['count', 'mean', 'median', 'max'])

In [255]:
gender_utilization

,count,mean,median,max
GENDER,,,,
FEMALE,6253,24.371662,11.0,605
MALE,6099,27.734711,9.0,825


In [256]:
gender_distribution = (
    patient_utilization
    .groupby('GENDER')['visit_count']
    .quantile([0.25, 0.50, 0.75, 0.90])
    .unstack()
)

gender_distribution

,0.25,0.50,0.75,0.90
GENDER,,,,
FEMALE,5.0,11.0,32.0,60.0
MALE,4.0,9.0,34.0,66.0


In [257]:
"""In this Synthea population, female patients had a slightly higher median encounter count, while male patients showed higher utilization in the upper part of the distribution."""

'In this Synthea population, female patients had a slightly higher median encounter count, while male patients showed higher utilization in the upper part of the distribution.'

In [258]:
encounters[['START', 'STOP']].head()

,START,STOP
0,2019-02-16T01:02:32Z,2019-02-16T01:17:32Z
1,2019-08-02T01:02:32Z,2019-08-02T01:32:32Z
2,2019-10-31T01:02:32Z,2019-10-31T01:17:32Z
3,2020-01-31T01:02:32Z,2020-01-31T01:17:32Z
4,2020-03-02T01:02:32Z,2020-03-02T01:58:32Z


In [259]:
encounters[['START', 'STOP']].dtypes

START    str
STOP     str
dtype: object

In [260]:
encounters['START']=pd.to_datetime(encounters['START'],errors='coerce')
encounters['STOP']=pd.to_datetime(encounters['STOP'],errors='coerce')

In [261]:
encounters[['START', 'STOP']].dtypes

START    datetime64[us, UTC]
STOP     datetime64[us, UTC]
dtype: object

In [262]:
encounters[['START', 'STOP']].head()

,START,STOP
0,2019-02-16 01:02:32+00:00,2019-02-16 01:17:32+00:00
1,2019-08-02 01:02:32+00:00,2019-08-02 01:32:32+00:00
2,2019-10-31 01:02:32+00:00,2019-10-31 01:17:32+00:00
3,2020-01-31 01:02:32+00:00,2020-01-31 01:17:32+00:00
4,2020-03-02 01:02:32+00:00,2020-03-02 01:58:32+00:00


In [263]:
"""Next concept: Extracting the year

Now we can finally answer:

How many healthcare encounters occurred each year?"""

'Next concept: Extracting the year\n\nNow we can finally answer:\n\nHow many healthcare encounters occurred each year?'

In [264]:
encounters['year']=encounters['START'].dt.year

In [265]:
encounters['year']

0         2019
1         2019
2         2019
3         2020
4         2020
          ... 
321523    2020
321524    2020
321525    2020
321526    2020
321527    2020
Name: year, Length: 321528, dtype: int32

In [266]:
encounters[['START', 'year']].head(10)

,START,year
0,2019-02-16 01:02:32+00:00,2019
1,2019-08-02 01:02:32+00:00,2019
2,2019-10-31 01:02:32+00:00,2019
3,2020-01-31 01:02:32+00:00,2020
4,2020-03-02 01:02:32+00:00,2020
5,2019-07-08 08:02:25+00:00,2019
6,2020-01-06 08:02:25+00:00,2020
7,2020-02-12 08:02:25+00:00,2020
8,2020-03-13 08:02:25+00:00,2020
9,2020-04-28 08:02:25+00:00,2020


In [267]:
#How many encounters occurred in each year?

In [268]:
encounters.groupby('year').size()

year
1910        4
1911        9
1912       18
1913       12
1914       13
        ...  
2016     7753
2017     7765
2018     8722
2019    25428
2020    25900
Length: 111, dtype: int64

In [269]:
yearly_utilization = (
    encounters
    .groupby('year')
    .size()
    .reset_index(name='encounter_count')
)

In [270]:
yearly_utilization

,year,encounter_count
0,1910,4
1,1911,9
2,1912,18
3,1913,12
4,1914,13
...,...,...
106,2016,7753
107,2017,7765
108,2018,8722
109,2019,25428


In [271]:
#By what percentage did encounter volume change compared with the previous year?
#Calculate year-over-year change(The method: .pct_change())

In [272]:
yearly_utilization['yoy_change']=(yearly_utilization['encounter_count'].pct_change()*100).round(2)

In [273]:
yearly_utilization

,year,encounter_count,yoy_change
0,1910,4,NaN
1,1911,9,125.00
2,1912,18,100.00
3,1913,12,-33.33
4,1914,13,8.33
...,...,...,...
106,2016,7753,4.36
107,2017,7765,0.15
108,2018,8722,12.32
109,2019,25428,191.54


In [274]:

yearly_utilization = yearly_utilization.sort_values('year')
yearly_utilization

,year,encounter_count,yoy_change
0,1910,4,NaN
1,1911,9,125.00
2,1912,18,100.00
3,1913,12,-33.33
4,1914,13,8.33
...,...,...,...
106,2016,7753,4.36
107,2017,7765,0.15
108,2018,8722,12.32
109,2019,25428,191.54


In [275]:
"""Average encounters per patient per year
   For every year we need:

Total encounters
÷
Number of unique patients
=
Average encounters per patient

For example:

2020:
50,000 encounters ÷ 5,000 patients
= 10 encounters/patient"""

'Average encounters per patient per year\n   For every year we need:\n\nTotal encounters\n÷\nNumber of unique patients\n=\nAverage encounters per patient\n\nFor example:\n\n2020:\n50,000 encounters ÷ 5,000 patients\n= 10 encounters/patient'

In [276]:
patients_per_year = (
    encounters
    .groupby('year')['PATIENT']
    .nunique()
    .reset_index(name='unique_patients')
)

In [277]:
patients_per_year

,year,unique_patients
0,1910,3
1,1911,6
2,1912,13
3,1913,7
4,1914,9
...,...,...
106,2016,4544
107,2017,4589
108,2018,5143
109,2019,8536


In [278]:
yearly_utilization = yearly_utilization.merge(
    patients_per_year,
    on='year',
    how='left'
)

In [279]:
yearly_utilization

,year,encounter_count,yoy_change,unique_patients
0,1910,4,NaN,3
1,1911,9,125.00,6
2,1912,18,100.00,13
3,1913,12,-33.33,7
4,1914,13,8.33,9
...,...,...,...,...
106,2016,7753,4.36,4544
107,2017,7765,0.15,4589
108,2018,8722,12.32,5143
109,2019,25428,191.54,8536


In [280]:
yearly_utilization['encounters_per_patient'] = (
    yearly_utilization['encounter_count']
    / yearly_utilization['unique_patients']
)

In [281]:
yearly_utilization

,year,encounter_count,yoy_change,unique_patients,encounters_per_patient
0,1910,4,NaN,3,1.333333
1,1911,9,125.00,6,1.500000
2,1912,18,100.00,13,1.384615
3,1913,12,-33.33,7,1.714286
4,1914,13,8.33,9,1.444444
...,...,...,...,...,...
106,2016,7753,4.36,4544,1.706206
107,2017,7765,0.15,4589,1.692090
108,2018,8722,12.32,5143,1.695897
109,2019,25428,191.54,8536,2.978913


In [282]:
high_utilization_threshold = visit_frequency['visit_count'].quantile(0.90)

high_utilization_threshold

np.float64(63.0)

In [283]:
""" So the 90th percentile of patient utilization is 63 visits.

What that means

Approximately:

90% of patients have 63 or fewer encounters
10% of patients have more than 63 encounters

This gives us a data-driven definition of a high-utilization patient."""

' So the 90th percentile of patient utilization is 63 visits.\n\nWhat that means\n\nApproximately:\n\n90% of patients have 63 or fewer encounters\n10% of patients have more than 63 encounters\n\nThis gives us a data-driven definition of a high-utilization patient.'

In [284]:
patient_utilization

,patient_id,visit_count,GENDER,age,age_group
0,0000b247-1def-417a-a783-41c8682be022,3,FEMALE,18,0-18
1,00049ee8-5953-4edd-a277-b9c1b1a7f16b,8,MALE,41,41-60
2,000769a6-23a7-426e-a264-cb0e509b2da2,4,FEMALE,79,60+
3,00079a57-24a8-430f-b4f8-a1cf34f90060,20,FEMALE,35,19-40
4,0008a63c-c95c-46c2-9ef3-831d68892019,5,MALE,33,19-40
...,...,...,...,...,...
12347,ffd3d544-1fcd-4a87-9514-fa6c37409cbc,75,MALE,76,60+
12348,ffd86fda-ebb9-400e-9fe3-ea1a1037dbad,14,FEMALE,34,19-40
12349,ffdbbb1b-745e-4e38-ade2-a19d6e778fee,34,MALE,67,60+
12350,ffdf0900-bc4b-4f81-b95b-1ea57da21e07,36,MALE,56,41-60


In [285]:
patient_utilization['high_utilizer'] = (
    patient_utilization['visit_count'] > high_utilization_threshold
)

In [286]:
patient_utilization

,patient_id,visit_count,GENDER,age,age_group,high_utilizer
0,0000b247-1def-417a-a783-41c8682be022,3,FEMALE,18,0-18,False
1,00049ee8-5953-4edd-a277-b9c1b1a7f16b,8,MALE,41,41-60,False
2,000769a6-23a7-426e-a264-cb0e509b2da2,4,FEMALE,79,60+,False
3,00079a57-24a8-430f-b4f8-a1cf34f90060,20,FEMALE,35,19-40,False
4,0008a63c-c95c-46c2-9ef3-831d68892019,5,MALE,33,19-40,False
...,...,...,...,...,...,...
12347,ffd3d544-1fcd-4a87-9514-fa6c37409cbc,75,MALE,76,60+,True
12348,ffd86fda-ebb9-400e-9fe3-ea1a1037dbad,14,FEMALE,34,19-40,False
12349,ffdbbb1b-745e-4e38-ade2-a19d6e778fee,34,MALE,67,60+,False
12350,ffdf0900-bc4b-4f81-b95b-1ea57da21e07,36,MALE,56,41-60,False


In [287]:
""" We're asking Python, for every patient:

Is this patient's visit count greater than 63?

For example:

visit_count = 20
20 > 63 → False


visit_count = 63
63 > 63 → False


visit_count = 100
100 > 63 → True  
 
 True → high-utilization patient
False → not high-utilization"""

" We're asking Python, for every patient:\n\nIs this patient's visit count greater than 63?\n\nFor example:\n\nvisit_count = 20\n20 > 63 → False\n\n\nvisit_count = 63\n63 > 63 → False\n\n\nvisit_count = 100\n100 > 63 → True  \n\n True → high-utilization patient\nFalse → not high-utilization"

In [288]:
#Check how many high utilizers we have

In [289]:
patient_utilization['high_utilizer'].value_counts()

high_utilizer
False    11153
True      1199
Name: count, dtype: int64

In [290]:
patient_utilization['high_utilizer'].value_counts(normalize=True)*100

high_utilizer
False    90.29307
True      9.70693
Name: proportion, dtype: float64

In [291]:
#What characteristics do these high-utilization patients have(1159 patients)?
#How many of our 1,199 high-utilization patients belong to each age group?

In [292]:
high_utilizer_age= (patient_utilization[patient_utilization['high_utilizer']].groupby('age_group').size())
high_utilizer_age

age_group
19-40      2
41-60     54
60+      829
dtype: int64

In [293]:
"""So 311 high-utilizers are missing from the age-group analysis."""

'So 311 high-utilizers are missing from the age-group analysis.'

In [294]:
patient_utilization = visit_frequency.merge(
    df_modified[['Id', 'GENDER', 'age', 'age_group']],
    left_on='patient_id',
    right_on='Id',
    how='left'
)

In [295]:
patient_utilization = patient_utilization.drop(columns='Id')

In [296]:
patient_utilization['high_utilizer'] = (
    patient_utilization['visit_count'] > high_utilization_threshold
)

In [297]:
patient_utilization[
    ['visit_count', 'GENDER', 'age', 'age_group', 'high_utilizer']
].isna().sum()

visit_count      0
GENDER           0
age              0
age_group        0
high_utilizer    0
dtype: int64

In [298]:
patient_utilization['age_group']

0          0-18
1         41-60
2        61-100
3         19-40
4         19-40
          ...  
12347    61-100
12348     19-40
12349    61-100
12350     41-60
12351     41-60
Name: age_group, Length: 12352, dtype: category
Categories (5, str): ['0-18' < '19-40' < '41-60' < '61-100' < '100+']

In [299]:
patient_utilization['high_utilizer'].value_counts()

high_utilizer
False    11153
True      1199
Name: count, dtype: int64

In [300]:
#What percentage of each age group is high-utilization?

In [301]:
age_high_utilization = (
    patient_utilization
    .groupby('age_group')['high_utilizer']
    .mean()
    * 100
)

age_high_utilization  #(For each age group, calculate the proportion of patients who are high-utilizers, then convert it to a percentage.)

age_group
0-18       0.000000
19-40      0.066050
41-60      1.883502
61-100    19.899184
100+      34.927697
Name: high_utilizer, dtype: float64

In [302]:
#So now we need to see both:

#Total patients in each age group
#High-utilizer patients in each age group
#High-utilizer percentage

In [303]:
"Create the full age-group profile"

'Create the full age-group profile'

In [304]:
age_profile =(patient_utilization.groupby('age_group').agg(
            total_patients=('high_utilizer','size'),
            high_utilizers=('high_utilizer','sum')
))
age_profile

,total_patients,high_utilizers
age_group,,
0-18,1392,0
19-40,3028,2
41-60,2867,54
61-100,4166,829
100+,899,314


In [305]:

"""Calculate the percentage ourselves

Once we have the two counts, we'll calculate:

high_utilizers
÷
total_patients
× 100"""

"Calculate the percentage ourselves\n\nOnce we have the two counts, we'll calculate:\n\nhigh_utilizers\n÷\ntotal_patients\n× 100"

In [306]:
age_profile['high_utilizer_pct'] = (
    age_profile['high_utilizers']
    / age_profile['total_patients']
    * 100
)

In [307]:
age_profile['high_utilizer_pct']=age_profile['high_utilizer_pct'].round(2)

In [308]:
age_profile  #last column is the pecentage of the hihhest utilizatio patiemts in that group

,total_patients,high_utilizers,high_utilizer_pct
age_group,,,
0-18,1392,0,0.00
19-40,3028,2,0.07
41-60,2867,54,1.88
61-100,4166,829,19.90
100+,899,314,34.93


In [309]:
"What proportion of male vs female patients are high-utilizers?"

'What proportion of male vs female patients are high-utilizers?'

In [310]:
gender_profile = (
    patient_utilization
    .groupby('GENDER')
    .agg(
        total_patients=('high_utilizer', 'size'),
        high_utilizers=('high_utilizer', 'sum')
    )
)

gender_profile

,total_patients,high_utilizers
GENDER,,
FEMALE,6253,548
MALE,6099,651


In [311]:
gender_profile['high_utilizer_pct'] = (
    gender_profile['high_utilizers']
    / gender_profile['total_patients']
    * 100
)

In [312]:
gender_profile['high_utilizer_pct'] = (
    gender_profile['high_utilizer_pct'].round(2)
)

In [313]:
gender_profile

,total_patients,high_utilizers,high_utilizer_pct
GENDER,,,
FEMALE,6253,548,8.76
MALE,6099,651,10.67


Does the gender difference remain within each age group?

In [314]:
age_gender_profile = (
    patient_utilization
    .groupby(['age_group', 'GENDER'])
    .agg(
        total_patients=('high_utilizer', 'size'),
        high_utilizers=('high_utilizer', 'sum')
    )
)

age_gender_profile

total_patients  high_utilizers
age_group GENDER                                
0-18      FEMALE             732               0
          MALE               660               0
19-40     FEMALE            1570               1
          MALE              1458               1
41-60     FEMALE            1528              32
          MALE              1339              22
61-100    FEMALE            2003             377
          MALE              2163             452
100+      FEMALE             420             138
          MALE               479             176

In [315]:
age_gender_profile['high_utilizer_pct'] = (
    age_gender_profile['high_utilizers']
    / age_gender_profile['total_patients']
    * 100
)

In [316]:
age_gender_profile['high_utilizer_pct'] = (
    age_gender_profile['high_utilizer_pct'].round(2)
)

In [317]:
age_gender_profile

total_patients  high_utilizers  high_utilizer_pct
age_group GENDER                                                   
0-18      FEMALE             732               0               0.00
          MALE               660               0               0.00
19-40     FEMALE            1570               1               0.06
          MALE              1458               1               0.07
41-60     FEMALE            1528              32               2.09
          MALE              1339              22               1.64
61-100    FEMALE            2003             377              18.82
          MALE              2163             452              20.90
100+      FEMALE             420             138              32.86
          MALE               479             176              36.74

Key analytical finding: high utilization increases dramatically with age. Within older groups, males have a higher high-utilizer proportion than females.

Quantify the gender difference

In [318]:
gender_age_comparison = (
    age_gender_profile
    .reset_index()
    .pivot(
        index='age_group',
        columns='GENDER',
        values='high_utilizer_pct'
    )
)

gender_age_comparison

GENDER,FEMALE,MALE
age_group,,
0-18,0.00,0.00
19-40,0.06,0.07
41-60,2.09,1.64
61-100,18.82,20.90
100+,32.86,36.74


In [319]:
gender_age_comparison['male_minus_female'] = (
    gender_age_comparison['MALE']
    - gender_age_comparison['FEMALE']
)

gender_age_comparison

GENDER,FEMALE,MALE,male_minus_female
age_group,,,
0-18,0.00,0.00,0.00
19-40,0.06,0.07,0.01
41-60,2.09,1.64,-0.45
61-100,18.82,20.90,2.08
100+,32.86,36.74,3.88


How to interpret it
Positive → males have a higher high-utilizer percentage.
Negative → females have a higher percentage.
0 → no difference.

High healthcare utilization is strongly concentrated in older age groups, with males showing a higher high-utilizer proportion than females among the older groups.

Find the strongest gender gap

In [320]:
gender_age_comparison.sort_values(
    'male_minus_female',
    ascending=False
)

GENDER,FEMALE,MALE,male_minus_female
age_group,,,
100+,32.86,36.74,3.88
61-100,18.82,20.90,2.08
19-40,0.06,0.07,0.01
0-18,0.00,0.00,0.00
41-60,2.09,1.64,-0.45


What types of encounters are driving high-utilizer patients?

In [323]:
encounters.head()

,Id,START,STOP,PATIENT,ORGANIZATION,PROVIDER,PAYER,ENCOUNTERCLASS,CODE,DESCRIPTION,BASE_ENCOUNTER_COST,TOTAL_CLAIM_COST,PAYER_COVERAGE,REASONCODE,REASONDESCRIPTION,year
0,d5ee30a9-362f-429e-a87a-ee38d999b0a5,2019-02-16 01:02:32+00:00,2019-02-16 01:17:32+00:00,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,5103c940-0c08-392f-95cd-446e0cea042a,e2c226c2-3e1e-3d0b-b997-ce9544c10528,7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a,outpatient,185345009,Encounter for symptom,129.16,129.16,69.16,65363002.0,Otitis media,2019
1,6a74fdef-2287-44bf-b9e7-18012376faca,2019-08-02 01:02:32+00:00,2019-08-02 01:32:32+00:00,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,0b9f3f7c-8ab6-30a5-b3ae-4dc0e0c00cb3,87c33fc5-3fd1-3c52-815a-b89a1623bb3a,7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a,wellness,410620009,Well child visit (procedure),129.16,129.16,129.16,NaN,NaN,2019
2,8bca6d8a-ab80-4cbf-8abb-46654235f227,2019-10-31 01:02:32+00:00,2019-10-31 01:17:32+00:00,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,5103c940-0c08-392f-95cd-446e0cea042a,e2c226c2-3e1e-3d0b-b997-ce9544c10528,7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a,outpatient,185345009,Encounter for symptom,129.16,129.16,69.16,65363002.0,Otitis media,2019
3,821e57ac-9304-46a9-9f9b-83daf60e9e43,2020-01-31 01:02:32+00:00,2020-01-31 01:17:32+00:00,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,0b9f3f7c-8ab6-30a5-b3ae-4dc0e0c00cb3,87c33fc5-3fd1-3c52-815a-b89a1623bb3a,7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a,wellness,410620009,Well child visit (procedure),129.16,129.16,129.16,NaN,NaN,2020
4,681c380b-3c84-4c55-80a6-db3d9ea12fee,2020-03-02 01:02:32+00:00,2020-03-02 01:58:32+00:00,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,fd328395-ab1d-35c6-a2d0-d05a9a79cf11,9c875a09-93e0-39aa-9260-ad264bbdd3fe,7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a,ambulatory,185345009,Encounter for symptom (procedure),129.16,129.16,69.16,NaN,NaN,2020


In [326]:
high_utilizer_ids = patient_utilization.loc[
    patient_utilization['high_utilizer'],
    'patient_id'
]

high_utilizer_ids.head()    #now have high_utilizer_ids, containing the patient IDs of your 1,199 high-utilizer patients.

5     00093cdd-a9f0-4ad8-87e9-53534501f008
15    002dd6c0-26c5-4bd0-b0f2-3b6f600b132f
32    007c77e6-8fc2-410f-adaf-62f12e8153d1
35    00848f6c-3abd-4963-b472-563dadb8ac7f
51    00c4a300-1dd2-40cb-a7c5-98d8dbb5e5d0
Name: patient_id, dtype: str

In [330]:
high_utilizer_encounters = encounters[
    encounters['PATIENT'].isin(high_utilizer_ids)
]

high_utilizer_encounters.shape

(145449, 16)

What types of encounters are most common among high-utilizer patients?

In [331]:
high_utilizer_encounters['ENCOUNTERCLASS'].value_counts()

ENCOUNTERCLASS
wellness      62176
outpatient    45715
ambulatory    29735
urgentcare     3427
inpatient      2468
emergency      1928
Name: count, dtype: int64

High utilization in this dataset is not necessarily driven by frequent hospitalization or emergency visits. A large portion comes from repeated routine/outpatient healthcare interactions.

Convert these counts into percentages

In [334]:
high_utilizer_encounter_pct = (
    high_utilizer_encounters['ENCOUNTERCLASS']
    .value_counts(normalize=True)
    .mul(100)
    .round(2)
)                                     #.value_counts(normalize=True)  #would give:

#wellness       0.50
#outpatient     0.30
#emergency      0.20       #.mul(100) means:

                           #multiply every value by 100
                                    

high_utilizer_encounter_pct

ENCOUNTERCLASS
wellness      42.75
outpatient    31.43
ambulatory    20.44
urgentcare     2.36
inpatient      1.70
emergency      1.33
Name: proportion, dtype: float64

In [335]:
high_utilizer_encounter_pct.sort_values(ascending=False)

ENCOUNTERCLASS
wellness      42.75
outpatient    31.43
ambulatory    20.44
urgentcare     2.36
inpatient      1.70
emergency      1.33
Name: proportion, dtype: float64

Which encounter types are disproportionately associated with high-utilizer patients?

We want to compare high-utilizer encounter distribution against the overall encounter distribution you calculated earlier.

In [338]:
overall_encounter_pct = (
    encounters['ENCOUNTERCLASS']
    .value_counts(normalize=True)
    .mul(100)
    .round(2)
)

encounter_comparison = pd.DataFrame({
    'overall_pct': overall_encounter_pct,
    'high_utilizer_pct': high_utilizer_encounter_pct
})

encounter_comparison['difference'] = (
    encounter_comparison['high_utilizer_pct']
    - encounter_comparison['overall_pct']
)

encounter_comparison.sort_values(                                 #This answers:

                                                                  #Which encounter types are more prominent among high-utilizer patients than in the overall population?
    'difference',
    ascending=False
)

,overall_pct,high_utilizer_pct,difference
ENCOUNTERCLASS,,,
outpatient,18.15,31.43,13.28
urgentcare,1.26,2.36,1.10
emergency,2.25,1.33,-0.92
inpatient,2.98,1.70,-1.28
wellness,48.59,42.75,-5.84
ambulatory,26.77,20.44,-6.33


Interpretation: High-utilizer patients have a substantially higher share of outpatient encounters than the overall population (31.43% vs 18.15%, +13.28 percentage points)

Identify the individual highest-utilization patients

In [339]:
top_10_utilizers = (
    patient_utilization[
        ['patient_id', 'visit_count', 'GENDER', 'age', 'age_group']
    ]
    .sort_values('visit_count', ascending=False)
    .head(10)
)

top_10_utilizers

,patient_id,visit_count,GENDER,age,age_group
12310,fef74b94-c2d9-4e22-bc4f-382c82152126,825,MALE,115,100+
4993,66f888db-3cd2-463e-8fa5-43c50dec2a3c,658,MALE,110,100+
8817,b7e10f32-477b-4a19-b6e4-5dca83d86574,611,MALE,106,100+
3779,4d9a197a-0469-4144-9030-0c04654cb4a8,605,FEMALE,102,100+
4156,55786035-203e-463f-87cf-3fc6cbf8b119,584,MALE,112,100+
7499,9cd159bb-f711-45b3-a353-1e354587b488,580,FEMALE,110,100+
12035,f93c04a1-b7b2-420d-9545-feab6d7dd1f0,544,MALE,85,61-100
10925,e396584a-3b3a-44e2-838f-be05097ef6bb,535,MALE,85,61-100
1845,2606458d-2648-49aa-af43-594a26be575f,525,MALE,104,100+
10456,d988bba1-094c-4bfa-a74f-befd2ad8c60b,519,MALE,111,100+


#High utilization is heavily concentrated among older patients.
#The high-utilizer proportion increases substantially with age.
#Among older groups, males generally have a higher high-utilizer proportion than females.
#Outpatient encounters are disproportionately represented among high-utilizer patients: 31.43% vs 18.15% overall.
#Emergency and inpatient encounters are relatively small components of high-utilizer activity.